crates/ecs/src/lib.rs

```diff
@@ -265,74 +259,293 @@ impl<'systems> Schedule<'systems> for Unordered<'systems> {
-                    Err(TryLockError::WouldBlock) => panic_locked_component_vec::<ComponentType>(),
-                    Err(TryLockError::Poisoned(_)) => panic!("Lock should not be poisoned!"),
-                };
-            }
+    fn create_new_entity(&mut self) -> Entity {
+        let entity_id = self.entities.len();
+        let entity = Entity {
+            id: entity_id,
+            _generation: 0,
+        };
+        self.entities.push(entity);
+        entity
+    }
+
+    fn add_empty_archetype(&mut self, archetype: Archetype) {
+        let archetype_index = self.archetypes.len();
+
+        archetype
+            .component_typeid_to_component_vec
+            .values()
+            .for_each(|v| {
+                let component_typeid = v.stored_type();
+                match self
+                    .component_typeid_to_archetype_indices
+                    .get_mut(&component_typeid)
+                {
+                    Some(indices) => indices.push(archetype_index),
+                    None => {
+                        self.component_typeid_to_archetype_indices
+                            .insert(component_typeid, vec![archetype_index]);
+                    }
+                }
+            });
+
+        self.archetypes.push(archetype);
+    }
+
+    fn store_entity_in_archetype(&mut self, entity_id: usize, archetype_index: usize) {
+        let archetype = self
+            .archetypes
+            .get_mut(archetype_index)
+            .expect("Archetype does not exist");
+
+        archetype.store_entity(entity_id);
+
+        if self.entity_id_to_archetype_index.contains_key(&entity_id) {
+            todo!("Add code for cleaning up old archetype");
         }
-        None
+
+        self.entity_id_to_archetype_index
+            .insert(entity_id, archetype_index);
     }
 
-    fn borrow_component_vec_mut<ComponentType: 'static>(&self) -> WriteComponentVec<ComponentType> {
-        let component_typeid = TypeId::of::<ComponentType>();
-        if let Some(component_vec) = self.component_vecs.get(&component_typeid) {
-            if let Some(component_vec) = component_vec
-                .as_any()
-                .downcast_ref::<ComponentVecImpl<ComponentType>>()
-            {
-                // This method should only be called once the scheduler has verified
-                // that component access can be done without contention.
-                // Panicking helps us detect errors in the scheduling algorithm more quickly.
-                return match component_vec.try_write() {
-                    Ok(component_vec) => Some(component_vec),
-                    Err(TryLockError::WouldBlock) => panic_locked_component_vec::<ComponentType>(),
-                    Err(TryLockError::Poisoned(_)) => panic!("Lock should not be poisoned!"),
-                };
-            }
+    fn add_component<ComponentType: Debug + Send + Sync + 'static>(
+        &mut self,
+        entity_id: usize,
```

Comment:
Because archetypes should not know about generation, they are only interested in entity Id. Adding Generation was necessary as it will not be needed until #53.  I've added a comment about this in the entity struct

Reply:
I see, but is it worth it to not use our specific `Entity` type whose purpose is to represent a unique entity? For sure, the generation might be unnecessary in some functions, but does it matter that it's present? 

I'd argue that it's both more consistent and more readable to always use `Entity`s and not break them up into their constituent parts (again, this goes back to the problem of using `usize` everywhere for different things). 

Or is there another reason not to simply pass in `Entity`, other than on principle of it not being necessary? A suggestion is to make the `id` field public and the `generation` field private, so code outside of the module isn't even aware of the `generation`.

Reply:
Perhaps using EntityId(usize) as its own type would make it easier to understand the HashMaps that use <usize, usize>

Reply:
Yeah, that'd be a lot better. 

But then what's the purpose of having two different entity types? `EntityId` and `Entity` represent the exact same thing, but one contains a smudge more information.

Reply:
Because you cannot use `Entity` in a HashMap, they will be impossible to find, thats why you want to map EntityId->ComponentIndex for example. But I think we can make these changes later, refactoring can happen in stages right?

Reply:
fixed 🔧🪛